### PR TITLE
merge lossgude and depthwise strategies 

### PR DESCRIPTION
**Pull request by ShvetsKS**

### CODE REVIEW

File: src/tree/updater_quantile_hist.h [hunk 1]
```diff
@@ -277,6 +277,68 @@ class QuantileHistMaker: public TreeUpdater {
       }
     };
 
+
+    // Drives execution of tree building on device
+    struct Driver {
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h:282]: I can help extract the driver into a more reusable module if needed.

Comment (ShvetsKS) [reply src/tree/updater_quantile_hist.h]: @trivialfis thanks for help, but seems I already did it!
Could you check the generalization of `Driver` for cpu and gpu updaters (common header file, methods for `ExpandEntry`, ...)?

File: src/tree/updater_quantile_hist.h [hunk 2]
```diff
@@ -148,6 +149,40 @@ struct CPUHistMakerTrainParam
   }
 };
 
+/* tree growing policies */
+struct ExpandEntry {
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h:153]: Do we need to worry about name colision?  Possibly during link time?

File: src/tree/updater_quantile_hist.h [hunk 3]
```diff
@@ -148,6 +149,40 @@ struct CPUHistMakerTrainParam
...
+  static const int kRootNid  = 0;
+  static const int kEmptyNid = -1;
+  int nid;
+  int sibling_nid;
+  int depth;
+  bst_float loss_chg;
+  unsigned timestamp;
+  ExpandEntry(int nid, int sibling_nid, int depth, bst_float loss_chg,
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h:161]: I think by now you have some familiarity with the GPU implementation, do you think it's feasible to get rid of the `sibling` in CPU implementation in the future?

File: src/tree/updater_quantile_hist.h [hunk 4]
```diff
@@ -365,85 +378,58 @@ class QuantileHistMaker: public TreeUpdater {
...
+                         int *num_leaves,
+                         unsigned *timestamp,
+                         std::vector<ExpandEntry>* nodes_for_apply_split);
+
     void BuildNodeStats(const GHistIndexMatrix &gmat,
-                        DMatrix *p_fmat,
-                        RegTree *p_tree,
-                        const std::vector<GradientPair> &gpair_h);
-
-    void EvaluateAndApplySplits(const GHistIndexMatrix &gmat,
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h:416]: I'm so thrilled to see remove functions!

**Review by trivialfis — COMMENTED** [review]
Thanks for working on this!  Just provided initial review, let me know when it's ready.

File: src/tree/updater_quantile_hist.h (hunk 1 above)
Comment (ShvetsKS) [reply src/tree/updater_quantile_hist.h]: @trivialfis  unfortunately I don't have complete understanding why previous and current commits are failed on gpu tests.
https://xgboost-ci.net/blue/organizations/jenkins/xgboost-win64/detail/PR-7007/6/pipeline
Continue to discover the exact reason of failed tests. 
without unification (462d1d3) it was ok :)

Comment (trivialfis) [reply src/tree/updater_quantile_hist.h]: Thanks, I will take a look tomorrow.

File: src/tree/updater_quantile_hist.h (hunk 2 above)
Comment (trivialfis) [reply src/tree/updater_quantile_hist.h:153]: @ShvetsKS I think this is a possible cause of the issue.

Comment (ShvetsKS) [reply src/tree/updater_quantile_hist.h:153]: @trivialfis  yes! Thanks for noticing it. I'll prepare fix

**Comment by trivialfis** [comment]
I wrote this patch, but seems I don't have permission to push to your branch:

``` patch
From a15d8a7d942ee7c8a582fec481571b76655c13c5 Mon Sep 17 00:00:00 2001
From: fis <jm.yuan@outlook.com>
Date: Tue, 1 Jun 2021 17:52:03 +0800
Subject: [PATCH] Hide the expand entry.

---
 src/tree/driver.h                      | 187 +++++++++++++------------
 src/tree/gpu_hist/expand_entry.cuh     |  84 +++--------
 src/tree/updater_gpu_hist.cu           |   3 +-
 src/tree/updater_quantile_hist.cc      |  12 +-
 src/tree/updater_quantile_hist.h       |  62 ++++----
 tests/cpp/tree/gpu_hist/test_driver.cu |   5 +-
 6 files changed, 149 insertions(+), 204 deletions(-)

diff --git a/src/tree/driver.h b/src/tree/driver.h
index faa55c84..9f25a58c 100644
--- a/src/tree/driver.h
+++ b/src/tree/driver.h
@@ -1,92 +1,95 @@
-/*!
- * Copyright 2021 by XGBoost Contributors
- */
-#ifndef XGBOOST_TREE_DRIVER_H_
-#define XGBOOST_TREE_DRIVER_H_
-#include <xgboost/span.h>
-#include <queue>
-#include <vector>
-#include "./param.h"
-
-namespace xgboost {
-namespace tree {
-
-template <typename ExpandEntryT>
-inline bool DepthWise(const ExpandEntryT& lhs, const ExpandEntryT& rhs) {
-  return lhs.GetDepth() > rhs.GetDepth();  // favor small depth
-}
-
-template <typename ExpandEntryT>
-inline bool LossGuide(const ExpandEntryT& lhs, const ExpandEntryT& rhs) {
-  if (lhs.GetLossChange() == rhs.GetLossChange()) {
-    return lhs.GetNodeId() > rhs.GetNodeId();  // favor small timestamp
-  } else {
-    return lhs.GetLossChange() < rhs.GetLossChange();  // favor large loss_chg
-  }
-}
-
-// Drives execution of tree building on device
-template <typename ExpandEntryT>
-class Driver {
-  using ExpandQueue =
-      std::priority_queue<ExpandEntryT, std::vector<ExpandEntryT>,
-                          std::function<bool(ExpandEntryT, ExpandEntryT)>>;
-
- public:
-  explicit Driver(TrainParam::TreeGrowPolicy policy)
-      : policy_(policy),
-        queue_(policy == TrainParam::kDepthWise ? DepthWise<ExpandEntryT> :
-                                                  LossGuide<ExpandEntryT>) {}
-  template <typename EntryIterT>
-  void Push(EntryIterT begin, EntryIterT end) {
-    for (auto it = begin; it != end; ++it) {
-      const ExpandEntryT& e = *it;
-      if (e.split.loss_chg > kRtEps) {
-        queue_.push(e);
-      }
-    }
-  }
-  void Push(const std::vector<ExpandEntryT> &entries) {
-    this->Push(entries.begin(), entries.end());
-  }
-  void Push(const ExpandEntryT e) {
-    queue_.push(e);
-  }
-
-  bool IsEmpty() {
-    return queue_.empty();
-  }
-
-  // Return the set of nodes to be expanded
-  // This set has no dependencies between entries so they may be expanded in
-  // parallel or asynchronously
-  std::vector<ExpandEntryT> Pop() {
-    if (queue_.empty()) return {};
-    // Return a single entry for loss guided mode
-    if (policy_ == TrainParam::kLossGuide) {
-      ExpandEntryT e = queue_.top();
-      queue_.pop();
-      return {e};
-    }
-    // Return nodes on same level for depth wise
-    std::vector<ExpandEntryT> result;
-    ExpandEntryT e = queue_.top();
-    int level = e.depth;
-    while (e.depth == level && !queue_.empty()) {
-      queue_.pop();
-      result.emplace_back(e);
-      if (!queue_.empty()) {
-        e = queue_.top();
-      }
-    }
-    return result;
-  }
-
- private:
-  TrainParam::TreeGrowPolicy policy_;
-  ExpandQueue queue_;
-};
-}  // namespace tree
-}  // namespace xgboost
-
-#endif  // XGBOOST_TREE_DRIVER_H_
+/*!
+ * Copyright 2021 by XGBoost Contributors
+ */
+#ifndef XGBOOST_TREE_DRIVER_H_
+#define XGBOOST_TREE_DRIVER_H_
+#include <xgboost/span.h>
+
+#include <functional>  // std::function
+#include <queue>
+#include <vector>
+
+#include "./param.h"
+
+namespace xgboost {
+namespace tree {
+
+template <typename ExpandEntryT>
+inline bool DepthWise(const ExpandEntryT& lhs, const ExpandEntryT& rhs) {
+  return lhs.GetDepth() > rhs.GetDepth();  // favor small depth
+}
+
+template <typename ExpandEntryT>
+inline bool LossGuide(const ExpandEntryT& lhs, const ExpandEntryT& rhs) {
+  if (lhs.GetLossChange() == rhs.GetLossChange()) {
+    return lhs.GetNodeId() > rhs.GetNodeId();  // favor small timestamp
+  } else {
+    return lhs.GetLossChange() < rhs.GetLossChange();  // favor large loss_chg
+  }
+}
+
+// Drives execution of tree building on device
+template <typename ExpandEntryT>
+class Driver {
+  using ExpandQueue =
+      std::priority_queue<ExpandEntryT, std::vector<ExpandEntryT>,
+                          std::function<bool(ExpandEntryT, ExpandEntryT)>>;
+
+ public:
+  explicit Driver(TrainParam::TreeGrowPolicy policy)
+      : policy_(policy),
+        queue_(policy == TrainParam::kDepthWise ? DepthWise<ExpandEntryT> :
+                                                  LossGuide<ExpandEntryT>) {}
+  template <typename EntryIterT>
+  void Push(EntryIterT begin, EntryIterT end) {
+    for (auto it = begin; it != end; ++it) {
+      const ExpandEntryT& e = *it;
+      if (e.split.loss_chg > kRtEps) {
+        queue_.push(e);
+      }
+    }
+  }
+  void Push(const std::vector<ExpandEntryT> &entries) {
+    this->Push(entries.begin(), entries.end());
+  }
+  void Push(const ExpandEntryT e) {
+    queue_.push(e);
+  }
+
+  bool IsEmpty() {
+    return queue_.empty();
+  }
+
+  // Return the set of nodes to be expanded
+  // This set has no dependencies between entries so they may be expanded in
+  // parallel or asynchronously
+  std::vector<ExpandEntryT> Pop() {
+    if (queue_.empty()) return {};
+    // Return a single entry for loss guided mode
+    if (policy_ == TrainParam::kLossGuide) {
+      ExpandEntryT e = queue_.top();
+      queue_.pop();
+      return {e};
+    }
+    // Return nodes on same level for depth wise
+    std::vector<ExpandEntryT> result;
+    ExpandEntryT e = queue_.top();
+    int level = e.depth;
+    while (e.depth == level && !queue_.empty()) {
+      queue_.pop();
+      result.emplace_back(e);
+      if (!queue_.empty()) {
+        e = queue_.top();
+      }
+    }
+    return result;
+  }
+
+ private:
+  TrainParam::TreeGrowPolicy policy_;
+  ExpandQueue queue_;
+};
+}  // namespace tree
+}  // namespace xgboost
+
+#endif  // XGBOOST_TREE_DRIVER_H_
diff --git a/src/tree/gpu_hist/expand_entry.cuh b/src/tree/gpu_hist/expand_entry.cuh
index 63216a27..b5a1f14c 100644
--- a/src/tree/gpu_hist/expand_entry.cuh
+++ b/src/tree/gpu_hist/expand_entry.cuh
@@ -1,11 +1,9 @@
 /*!
  * Copyright 2020 by XGBoost Contributors
  */
-#ifndef DRIVER_CUH_
-#define DRIVER_CUH_
+#ifndef EXPAND_ENTRY_CUH_
+#define EXPAND_ENTRY_CUH_
 #include <xgboost/span.h>
-#include <queue>
-#include "../param.h"
 #include "evaluate_splits.cuh"
 
 namespace xgboost {
@@ -47,6 +45,18 @@ struct ExpandEntry {
     return true;
   }
 
+  bst_float GetLossChange() const {
+    return split.loss_chg;
+  }
+
+  int GetNodeId() const {
+    return nid;
+  }
+
+  int GetDepth() const {
+    return depth;
+  }
+
   friend std::ostream& operator<<(std::ostream& os, const ExpandEntry& e) {
     os << "ExpandEntry: \n";
     os << "nidx: " << e.nid << "\n";
@@ -56,72 +66,10 @@ struct ExpandEntry {
     os << "right_sum: " << e.split.right_sum << "\n";
     return os;
   }
-};
-
-inline bool DepthWise(const ExpandEntry& lhs, const ExpandEntry& rhs) {
-  return lhs.depth > rhs.depth;  // favor small depth
-}
-
-inline bool LossGuide(const ExpandEntry& lhs, const ExpandEntry& rhs) {
-  if (lhs.split.loss_chg == rhs.split.loss_chg) {
-    return lhs.nid > rhs.nid;  // favor small timestamp
-  } else {
-    return lhs.split.loss_chg < rhs.split.loss_chg;  // favor large loss_chg
-  }
-}
 
-// Drives execution of tree building on device
-class Driver {
-  using ExpandQueue =
-      std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
-                          std::function<bool(ExpandEntry, ExpandEntry)>>;
-
- public:
-  explicit Driver(TrainParam::TreeGrowPolicy policy)
-      : policy_(policy),
-        queue_(policy == TrainParam::kDepthWise ? DepthWise : LossGuide) {}
-  template <typename EntryIterT>
-  void Push(EntryIterT begin,EntryIterT end) {
-    for (auto it = begin; it != end; ++it) {
-      const ExpandEntry& e = *it;
-      if (e.split.loss_chg > kRtEps) {
-        queue_.push(e);
-      }
-    }
-  }
-  void Push(const std::vector<ExpandEntry> &entries) {
-    this->Push(entries.begin(), entries.end());
-  }
-  // Return the set of nodes to be expanded
-  // This set has no dependencies between entries so they may be expanded in
-  // parallel or asynchronously
-  std::vector<ExpandEntry> Pop() {
-    if (queue_.empty()) return {};
-    // Return a single entry for loss guided mode
-    if (policy_ == TrainParam::kLossGuide) {
-      ExpandEntry e = queue_.top();
-      queue_.pop();
-      return {e};
-    }
-    // Return nodes on same level for depth wise
-    std::vector<ExpandEntry> result;
-    ExpandEntry e = queue_.top();
-    int level = e.depth;
-    while (e.depth == level && !queue_.empty()) {
-      queue_.pop();
-      result.emplace_back(e);
-      if (!queue_.empty()) {
-        e = queue_.top();
-      }
-    }
-    return result;
-  }
-
- private:
-  TrainParam::TreeGrowPolicy policy_;
-  ExpandQueue queue_;
 };
+
 }  // namespace tree
 }  // namespace xgboost
 
-#endif  // DRIVER_CUH_
\ No newline at end of file
+#endif  // EXPAND_ENTRY_CUH_
diff --git a/src/tree/updater_gpu_hist.cu b/src/tree/updater_gpu_hist.cu
index 83c197f3..7919e22b 100644
--- a/src/tree/updater_gpu_hist.cu
+++ b/src/tree/updater_gpu_hist.cu
@@ -25,6 +25,7 @@
 #include "../data/ellpack_page.cuh"
 
 #include "param.h"
+#include "driver.h"
 #include "updater_gpu_common.cuh"
 #include "split_evaluator.h"
 #include "constraints.cuh"
@@ -694,7 +695,7 @@ struct GPUHistMakerDevice {
   void UpdateTree(HostDeviceVector<GradientPair>* gpair_all, DMatrix* p_fmat,
                   RegTree* p_tree, dh::AllReducer* reducer) {
     auto& tree = *p_tree;
-    Driver driver(static_cast<TrainParam::TreeGrowPolicy>(param.grow_policy));
+    Driver<ExpandEntry> driver(static_cast<TrainParam::TreeGrowPolicy>(param.grow_policy));
 
     monitor.Start("Reset");
     this->Reset(gpair_all, p_fmat, p_fmat->Info().num_col_);
diff --git a/src/tree/updater_quantile_hist.cc b/src/tree/updater_quantile_hist.cc
index a160f15f..8e96713b 100644
--- a/src/tree/updater_quantile_hist.cc
+++ b/src/tree/updater_quantile_hist.cc
@@ -144,7 +144,7 @@ void BatchHistSynchronizer<GradientSumT>::SyncHistograms(BuilderT *builder,
       const size_t parent_id = (*p_tree)[entry.nid].Parent();
       auto parent_hist = builder->hist_[parent_id];
       auto sibling_hist = builder->hist_[entry.sibling_nid];
-      SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
+      common::SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
     }
   });
   builder->builder_monitor_.Stop("SyncHistograms");
@@ -173,7 +173,7 @@ void DistributedHistSynchronizer<GradientSumT>::SyncHistograms(BuilderT* builder
       const size_t parent_id = (*p_tree)[entry.nid].Parent();
       auto parent_hist = builder->hist_local_worker_[parent_id];
       auto sibling_hist = builder->hist_[entry.sibling_nid];
-      SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
+      common::SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
       // Store posible parent node
       auto sibling_local = builder->hist_local_worker_[entry.sibling_nid];
       CopyHist(sibling_local, sibling_hist, r.begin(), r.end());
@@ -199,7 +199,7 @@ template <typename GradientSumT>
 void DistributedHistSynchronizer<GradientSumT>::ParallelSubtractionHist(
                                   BuilderT* builder,
                                   const common::BlockedSpace2d& space,
-                                  const std::vector<ExpandEntry>& nodes,
+                                  const std::vector<QuantileHistMaker::ExpandEntry>& nodes,
                                   const RegTree * p_tree) {
   common::ParallelFor2d(space, builder->nthread_, [&](size_t node, common::Range1d r) {
     const auto& entry = nodes[node];
@@ -209,7 +209,8 @@ void DistributedHistSynchronizer<GradientSumT>::ParallelSubtractionHist(
       if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1) {
         auto parent_hist = builder->hist_[(*p_tree)[entry.nid].Parent()];
         auto sibling_hist = builder->hist_[entry.sibling_nid];
-        SubtractionHist(this_hist, parent_hist, sibling_hist, r.begin(), r.end());
+        common::SubtractionHist(this_hist, parent_hist, sibling_hist, r.begin(),
+                                r.end());
       }
     }
   });
@@ -431,8 +432,7 @@ void QuantileHistMaker::Builder<GradientSumT>::BuildNodeStats(
   const DMatrix& fmat,
   const std::vector<GradientPair> &gpair_h,
   const std::vector<ExpandEntry>& nodes_for_apply_split, RegTree *p_tree) {
-  for (size_t i = 0; i < nodes_for_apply_split.size(); ++i) {
-    const ExpandEntry candidate = nodes_for_apply_split[i];
+  for (auto const& candidate : nodes_for_apply_split) {
     const int nid = candidate.nid;
     const int cleft = (*p_tree)[nid].LeftChild();
     const int cright = (*p_tree)[nid].RightChild();
diff --git a/src/tree/updater_quantile_hist.h b/src/tree/updater_quantile_hist.h
index 4ecb7fb9..4d6ad7ed 100644
--- a/src/tree/updater_quantile_hist.h
+++ b/src/tree/updater_quantile_hist.h
@@ -149,42 +149,36 @@ struct CPUHistMakerTrainParam
   }
 };
 
-/* tree growing policies */
-struct ExpandEntry {
-  static const int kRootNid  = 0;
-  static const int kEmptyNid = -1;
-  int nid;
-  int sibling_nid;
-  int depth;
-  bst_float loss_chg;
-  unsigned timestamp;
-  ExpandEntry(int nid, int sibling_nid, int depth, bst_float loss_chg,
-              unsigned tstmp)
-      : nid(nid), sibling_nid(sibling_nid), depth(depth),
-        loss_chg(loss_chg), timestamp(tstmp) {}
-
-  bool IsValid(TrainParam const &param, int32_t num_leaves) const {
-    bool ret = loss_chg <= kRtEps ||
-               (param.max_depth > 0 && this->depth == param.max_depth) ||
-               (param.max_leaves > 0 && num_leaves == param.max_leaves);
-    return ret;
-  }
+/*! \brief construct a tree using quantized feature values */
+class QuantileHistMaker: public TreeUpdater {
+ public:
+  struct ExpandEntry {
+    static const int kRootNid = 0;
+    static const int kEmptyNid = -1;
+    int nid;
+    int sibling_nid;
+    int depth;
+    bst_float loss_chg;
+    unsigned timestamp;
+    ExpandEntry(int nid, int sibling_nid, int depth, bst_float loss_chg,
+                unsigned tstmp)
+        : nid(nid), sibling_nid(sibling_nid), depth(depth), loss_chg(loss_chg),
+          timestamp(tstmp) {}
+
+    bool IsValid(TrainParam const &param, int32_t num_leaves) const {
+      bool ret = loss_chg <= kRtEps ||
+                 (param.max_depth > 0 && this->depth == param.max_depth) ||
+                 (param.max_leaves > 0 && num_leaves == param.max_leaves);
+      return ret;
+    }
 
-  bst_float GetLossChange() const {
-    return loss_chg;
-  }
+    bst_float GetLossChange() const { return loss_chg; }
 
-  int GetNodeId() const {
-    return nid;
-  }
+    int GetNodeId() const { return nid; }
 
-  int GetDepth() const {
-    return depth;
-  }
-};
+    int GetDepth() const { return depth; }
+  };
 
-/*! \brief construct a tree using quantized feature values */
-class QuantileHistMaker: public TreeUpdater {
  public:
   QuantileHistMaker() {
     updater_monitor_.Init("QuantileHistMaker");
@@ -539,7 +533,7 @@ class DistributedHistSynchronizer: public HistSynchronizer<GradientSumT> {
 
   void ParallelSubtractionHist(BuilderT* builder,
                                const common::BlockedSpace2d& space,
-                               const std::vector<ExpandEntry>& nodes,
+                               const std::vector<QuantileHistMaker::ExpandEntry>& nodes,
                                const RegTree * p_tree);
 };
 
@@ -568,8 +562,6 @@ class DistributedHistRowsAdder: public HistRowsAdder<GradientSumT> {
   void AddHistRows(BuilderT*, int *starting_index,
                    int *sync_count, RegTree *p_tree) override;
 };
-
-
 }  // namespace tree
 }  // namespace xgboost
 
diff --git a/tests/cpp/tree/gpu_hist/test_driver.cu b/tests/cpp/tree/gpu_hist/test_driver.cu
index 488855b4..40be10ba 100644
--- a/tests/cpp/tree/gpu_hist/test_driver.cu
+++ b/tests/cpp/tree/gpu_hist/test_driver.cu
@@ -1,11 +1,12 @@
 #include <gtest/gtest.h>
 #include "../../../../src/tree/gpu_hist/expand_entry.cuh"
+#include "../../../../src/tree/driver.h"
 
 namespace xgboost {
 namespace tree {
 
 TEST(GpuHist, DriverDepthWise) {
-  Driver driver(TrainParam::kDepthWise);
+  Driver<ExpandEntry> driver(TrainParam::kDepthWise);
   EXPECT_TRUE(driver.Pop().empty());
   DeviceSplitCandidate split;
   split.loss_chg = 1.0f;
@@ -32,7 +33,7 @@ TEST(GpuHist, DriverLossGuided) {
   DeviceSplitCandidate low_gain;
   low_gain.loss_chg = 1.0f;
 
-  Driver driver(TrainParam::kLossGuide);
+  Driver<ExpandEntry> driver(TrainParam::kLossGuide);
   EXPECT_TRUE(driver.Pop().empty());
   ExpandEntry root(0, 0, high_gain, .0f, .0f, .0f);
   driver.Push({root});
-- 
2.31.1

```

File: src/tree/updater_quantile_hist.h (hunk 3 above)
Comment (ShvetsKS) [reply src/tree/updater_quantile_hist.h]: sure. I tried to avoid it. Please check it

**Comment by ShvetsKS** [comment]
@trivialfis thanks a lot for provided solution!
I applied similar ones right after I've managed to run tests on gpu and debug it thanks to yours notice.

I was trying to merge `ExpandEntry` for cpu and gpu code but it requires too much changes in both branches, seems it can be done separately if needed. So at this time different naming is provided. If renaming is not applicable I'll apply your patch.

About write access to branch, it was enabled, feel free to modify it :)

**Comment by trivialfis** [comment]
I stopped the jvm test as it's hanging in:

> [2021-06-01T19:15:09.094Z] - test with quantile hist with interaction_constraints (depthwise) *** FAILED ***

**Comment by ShvetsKS** [comment]
@trivialfis do you think hanging jvm tests are related to current changes?

**Comment by trivialfis** [comment]
Possible, since both tests in github action and jenkins are failing, the one on github action is:

>  - test with quantile hist with monotone_constraints (lossguide) *** FAILED ***

So both of them are lossguide hist.

**Comment by ShvetsKS** [comment]
> Possible, since both tests in github action and jenkins are failing, the one on github action is:
> 
> > * test with quantile hist with monotone_constraints (lossguide) *** FAILED ***
> 
> So both of them are lossguide hist.

seems the fail cause is removing `sibling_nid`. I'll try to fix it

**Comment by trivialfis** [comment]
No problem, let me know if there's anything I can help.

**Comment by ShvetsKS** [comment]
> No problem, let me know if there's anything I can help.

seems all problems were resolved

**Comment by trivialfis** [comment]
Awesome!  Is it still WIP?

**Comment by ShvetsKS** [comment]
> ~Awesome! Is it still WIP?~ Just saw the changed titile.

yes. Seems it's ready for review :)

**Review by trivialfis — APPROVED** [review]

**Pull request merged**